execution/block-partitioner/src/sharded_block_partitioner/partitioning_shard.rs

```diff
@@ -0,0 +1,242 @@
+            match msg {
+                CrossShardMsg::AcceptedTxnsMsg(num_accepted_txns) => {
+                    accepted_txns_vec[i] = num_accepted_txns;
+                },
+                _ => panic!("Unexpected message"),
+            }
+        }
+        accepted_txns_vec
+    }
+
+    fn discard_txns_with_cross_shard_deps(&self, partition_msg: DiscardTxnsWithCrossShardDep) {
+        let DiscardTxnsWithCrossShardDep {
+            transactions,
+            prev_rounds_rw_set_with_index,
+            prev_rounds_frozen_chunks,
+        } = partition_msg;
+        let num_shards = self.message_txs.len();
+        let mut conflict_detector = CrossShardConflictDetector::new(self.shard_id, num_shards);
+        // If transaction filtering is allowed, we need to prepare the dependency analysis and broadcast it to other shards
+        // Based on the dependency analysis received from other shards, we will reject transactions that are conflicting with
+        // transactions in other shards
+        let read_write_set = RWSet::new(&transactions);
+        self.broadcast_rw_set(read_write_set);
+        let cross_shard_rw_set = self.collect_rw_set();
+        let (accepted_txns, accepted_cross_shard_dependencies, rejected_txns) = conflict_detector
+            .discard_txns_with_cross_shard_deps(
+                transactions,
+                &cross_shard_rw_set,
+                prev_rounds_rw_set_with_index,
+            );
+        // Broadcast and collect the stats around number of accepted and rejected transactions from other shards
+        // this will be used to determine the absolute index of accepted transactions in this shard.
+        self.broadcast_num_accepted_txns(accepted_txns.len());
+        let accepted_txns_vec = self.collect_num_accepted_txns();
+        // Calculate the absolute index of accepted transactions in this shard, which is the sum of all accepted transactions
+        // from other shards whose shard id is smaller than the current shard id and the number of accepted transactions in the
+        // previous rounds
+        let mut index_offset = prev_rounds_frozen_chunks
+            .iter()
+            .map(|chunk| chunk.len())
+            .sum::<usize>();
+        for num_accepted_txns in accepted_txns_vec.iter().take(self.shard_id) {
+            index_offset += num_accepted_txns;
+        }
+
+        // Calculate the RWSetWithTxnIndex for the accepted transactions
+        let current_rw_set_with_index = RWSetWithTxnIndex::new(&accepted_txns, index_offset);
+
+        let accepted_txns_with_dependencies = accepted_txns
+            .into_iter()
+            .zip(accepted_cross_shard_dependencies.into_iter())
+            .map(|(txn, dependencies)| TransactionWithDependencies::new(txn, dependencies))
+            .collect::<Vec<TransactionWithDependencies>>();
+
+        let frozen_chunk = SubBlock::new(index_offset, accepted_txns_with_dependencies);
+        drop(prev_rounds_frozen_chunks);
```

Comment:
This is needed because, the reference needs to be dropped explicitly before sending the response back as on the other side, we rely on unwrapping the Arc.

Reply:
Oh I see. Document it. (Also looks like we can just pass in a number instead of this `Arc`?)

Reply:
Documented, I am passing `Arc` instead of number because to avoid the additional copy.